src/main/java/net/trollyloki/advancementbingo/BingoManager.java

```diff
@@ -111,8 +119,87 @@ public void onWin(@NotNull BingoTeam team) {
+        int col = advancementLocation[1];
+
+        // Hit for {Team name}
+        // Progress on rows/columns/diagonals
+        // Total bingos scored
+
+        Component hitMessage = hitMessage(team);
+        Component lineMessage = progressMessage(board, row, col);
+        Component completedMessage = Component.text("Completed " + board.getCompletedRows() + "/" + board.getRequiredRows() + " bingos to win")
+                .color(team.getTextColor());
+
+        Bukkit.getScheduler().runTask(plugin, () -> {
+            for (UUID uuid : players.keySet()) {
+                Player player = Bukkit.getPlayer(uuid);
+                if (player == null) continue;
+                player.sendMessage(hitMessage);
+                player.sendMessage(lineMessage);
+                if (board.getCompletedRows() > 0 && board.getRequiredRows() > 1) player.sendMessage(completedMessage);
+                Jingles.playAdvancementHit(plugin, player);
+            }
+        });
+
+    }
+
+    private void addOrAppend(Map<Integer, String> map, int key, String value) {
+        if (map.containsKey(key)) {
+            map.put(key, map.get(key) + ", " + value);
+        } else {
+            map.put(key, value);
+        }
+
+    }
+
+    private Component hitMessage(BingoTeam team) {
+        return Component.text().content("Hit for ")
+                .append(Component.text("&")
+                        .color(team.getTextColor())
+                        .decoration(TextDecoration.OBFUSCATED, true))
+                .append(team.getDisplayName().color(team.getTextColor()))
+                .append(Component.text("&")
+                        .color(team.getTextColor())
+                        .decoration(TextDecoration.OBFUSCATED, true))
+                .decoration(TextDecoration.BOLD, true)
+                .build();
+    }
+
+    private Component progressMessage(BingoBoard board, int row, int col) {
+        HashMap<Integer, String> lineProgress = new HashMap<>();
+
+        addOrAppend(lineProgress, board.getRowProgress(row), "Row " + (row+1));
+        addOrAppend(lineProgress, board.getColumnProgress(col), "Column " + (col+1));
+        if (row == col) addOrAppend(lineProgress, board.getTopLeftDiagonalProgress(), "Left Diagonal");
+        if (row == board.getLength() - 1 - col) addOrAppend(lineProgress, board.getTopRightDiagonalProgress(), "Right Diagonal");
+
+        List<TextComponent> components = new ArrayList<>();
+        lineProgress.keySet().stream()
+                .sorted(Comparator.reverseOrder())
+                .map((key) -> {
+                    TextComponent.Builder builder = Component.text().content(key + "/" + board.getLength());
+                    if (key == board.getLength()) builder.color(NamedTextColor.GREEN);
+                    else builder.color(NamedTextColor.GRAY);
+                    builder.append(Component.text(" for " + lineProgress.get(key)).color(NamedTextColor.GRAY));
```

Comment:
It might be better to stream the entrySet() since you're always using the corresponding value anyway.